devito/ir/iet/nodes.py

```diff
@@ -273,51 +261,55 @@ def __repr__(self):
         return "%sCall::\n\t%s(...)" % (ret, self.name)
 
     @property
+    def children(self):
+        return tuple(i for i in self.arguments if isinstance(i, (Call, Lambda)))
+
+    @cached_property
     def functions(self):
         retval = []
         for i in self.arguments:
-            if isinstance(i, numbers.Number):
-                continue
-            elif isinstance(i, (AbstractFunction, Indexed, LocalObject)):
+            if isinstance(i, (AbstractFunction, Indexed, LocalObject)):
                 retval.append(i.function)
+            elif isinstance(i, Call):
+                retval.extend(i.functions)
             else:
-                for s in i.free_symbols:
+                try:
+                    v = i.free_symbols
+                except AttributeError:
+                    continue
+                for s in v:
                     try:
-                        f = s.function
+                        # `try-except` necessary for e.g. Macro
+                        if isinstance(s.function, AbstractFunction):
+                            retval.append(s.function)
                     except AttributeError:
                         continue
-                    if isinstance(f, AbstractFunction):
-                        retval.append(f)
         if self.base is not None:
             retval.append(self.base.function)
         if self.retobj is not None:
             retval.append(self.retobj.function)
         return tuple(filter_ordered(retval))
 
-    @property
-    def children(self):
-        return tuple(i for i in self.arguments if isinstance(i, (Call, Lambda)))
-
     @cached_property
-    def free_symbols(self):
-        free = set()
+    def expr_symbols(self):
+        retval = []
         for i in self.arguments:
-            if isinstance(i, numbers.Number):
+            if isinstance(i, AbstractFunction):
                 continue
-            elif isinstance(i, AbstractFunction):
-                if i.is_ArrayBasic:
-                    free.add(i)
-                else:
-                    # Always passed by _C_name since what actually needs to be
-                    # provided is the pointer to the corresponding C struct
-                    free.add(i._C_symbol)
+            elif isinstance(i, (Indexed, LocalObject, Symbol)):
+                retval.append(i)
+            elif isinstance(i, Call):
+                retval.extend(i.expr_symbols)
             else:
-                free.update(i.free_symbols)
+                try:
+                    retval.extend(i.free_symbols)
+                except AttributeError:
+                    pass
         if self.base is not None:
-            free.add(self.base)
+            retval.append(self.base)
         if self.retobj is not None:
-            free.update(self.retobj.free_symbols)
-        return free
+            retval.extend(self.retobj.free_symbols)
+        return tuple(filter_ordered(retval))
```

Comment:
maybe be good at some point to have the `filter` function to be split into different output type because this does a lot of `list(tuple(np(tuple)` which isn't great.

Reply:
yeah agree, though the overhead should be minimal